File: Daily_Challenge/20230304_LC_1345.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        // BFS solution
        // build graph
        unordered_map<int, vector<int>> graph;
        for (int i = 0; i < arr.size(); i++){
            graph[arr[i]].push_back(i);
        }
        unordered_set<int> seen;
        queue<int> q;
        q.push(0);
        seen.insert(0);
        int steps = 0;
        int n = arr.size();
        while (!q.empty()){
            int size = q.size();
            for (int iter = 0; iter < size; iter++){
                int cur = q.front();
                q.pop();
                //cout << cur << endl;
                if (cur == n - 1){
                    return steps;
                }
                // explore cur-1 and cur+1
                vector<int> nei = {cur-1, cur+1};
                for (int nxt: nei){
                    if (nxt >= 0 && nxt < n && !seen.count(nxt)){
                        q.push(nxt);
                        seen.insert(nxt);
                    }
                }
                // explore the same value indices
                for (int nxt: graph[arr[cur]]){
                    //cout << nxt << endl;
                    if (!seen.count(nxt)){
                        q.push(nxt);
                        seen.insert(nxt);
                    }
                }
                graph[arr[cur]].clear();
            }
            steps++;
        }

        return -1;
    }
};

File: Daily_Challenge/20230304_LC_1345.py
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        
        graph = collections.defaultdict(list)
        for idx, val in enumerate(arr):
            graph[val].append(idx)
        
        q = collections.deque()
        q.append(0)
        seen = set()
        seen.add(0)
        
        steps = 0
        
        n = len(arr)
        while q:
            size = len(q)
            for _ in range(size):
                cur_idx = q.popleft()
                if cur_idx == n-1:
                    return steps
                for nxt in [cur_idx-1, cur_idx+1]:
                    if 0 <= nxt < n and nxt not in seen:
                        q.append(nxt)
                        seen.add(nxt)
                for nxt in graph[arr[cur_idx]]:
                    if nxt not in seen:
                        q.append(nxt)
                        seen.add(nxt)
                graph[arr[cur_idx]].clear()
            steps += 1
            
        
        
        
        return -1